drivers/mma7660: Add initial support

immesys opened the pull request:
I am going through my branches looking for things to contribute back upstream, this is a driver for one of the cheapest accelerometers on the market ($0.80 each). The documentation is maybe not the friendliest, but I've been using the basic functionality of this driver for a while and it works well. I haven't used the tap detection feature, but to my knowledge, the code is correct.

Feedback welcome!

immesys commented:
Any feedback? I'll rebase this quickly

haukepetersen commented on drivers/include/mma7660.h:
```diff
@@ -0,0 +1,200 @@
+ * @return                   0 on success
+ * @return                  -1 on error
+ */
+int mma7660_init(mma7660_t *dev, i2c_t i2c, uint8_t address);
```
please change this to 
```c
int mma7660_init(mma7660_t *dev, const mma7660_params_t *params);
```
as done for e.g. the bh1750fvi or the dht drivers. For this also please add a default `mma7660_params.h` file, which will ready this driver to be used with auto_init etc.

haukepetersen commented on drivers/include/mma7660.h:
```diff
@@ -0,0 +1,200 @@
+ * @return                   0 on success
+ * @return                  -1 on error
+ */
+int mma7660_config_samplerate(mma7660_t *dev, uint8_t amsr, uint8_t awsr, uint8_t filt);
```
IMHO, this should also be configurable during initalization through the `mma7660_params_t` struct.

miri64 replied:
@immesys please address

miri64 replied:
@haukepetersen could you clarify?

haukepetersen replied:
seems like this was addressed already.

haukepetersen replied:
hm, only partly, see comments on the `_params.h` file

haukepetersen commented on drivers/include/mma7660.h:
```diff
@@ -0,0 +1,200 @@
+ * @return                   0 on success
+ * @return                  -1 on error
+ */
+int mma7660_read(mma7660_t *dev, int8_t *x, int8_t *y, int8_t *z);
```
is the accelleromter only sampling with 8-bit?

In any case, I would strongly prefer to output the accel values in their physical representation in mili G [mg]:
```c
void mma7660_read(mma7660_t *dev, mma7660_data_t *data);

// with
typedef struct {
    int16_t x;
    int16_t y;
    int16_t z;
} mma7660_data_t;
```

haukepetersen commented:
@immesys ping?!

immesys commented:
Sorry, my available time didn't line up before the release. I'll be getting back to our RIOT projects in a week or two

immesys commented:
Ok I rebased the code and did the refactoring that you requested.

miri64 commented on drivers/mma7660/mma7660.c:
```diff
@@ -0,0 +1,209 @@
+}
+
+int mma7660_read(mma7660_t *dev, int16_t *x, int16_t *y, int16_t *z) {
+  int8_t countx, county, countz;
```
Indentation

miri64 commented on drivers/mma7660/include/mma7660_reg.h:
```diff
@@ -0,0 +1,49 @@
+#define MMA7660_PD      0x0A /**< Tap / Pulse debounce */
+
+
+
```
Any reasons for this many spaces?

miri64 commented on drivers/mma7660/include/mma7660_params.h:
```diff
@@ -0,0 +1,64 @@
+#define MMA7660_PARAM_ADDR        (MMA7660_ADDR)
+#endif
+
+#define MMA7660_PARAMS_DEFAULT    {.i2c = MMA7660_PARAM_I2C, \
```
Doc missing

miri64 commented on drivers/mma7660/include/mma7660_params.h:
```diff
@@ -0,0 +1,64 @@
+#define MMA7660_PARAM_I2C         I2C_DEV(0)
+#endif
+#ifndef MMA7660_PARAM_ADDR
+#define MMA7660_PARAM_ADDR        (MMA7660_ADDR)
```
Doc missing

immesys commented:
Rebased onto new release, fixed most of the outstanding comments. I wasn't sure what extra documentation you wanted. I took a look at a few other sensors' params.h files and they didn't have it either so I didn't know what was missing. Any pointers?

miri64 commented:
>  I wasn't sure what extra documentation you wanted. I took a look at a few other sensors' params.h files and they didn't have it either so I didn't know what was missing. Any pointers?

Mh curious, usually the docchecker should pick up something like that. I just wanted to prevent that for your PR. Just a `@brief The default configuration for MMA7760` above the macro definition should suffice.

haukepetersen commented on drivers/mma7660/include/mma7660_params.h:
```diff
@@ -0,0 +1,64 @@
+#ifndef MMA7660_PARAM_ADDR
+#define MMA7660_PARAM_ADDR        (MMA7660_ADDR)
+#endif
+
```
Default values for `amsr`, `awsr`, and `filt` are missing.

immesys commented:
@haukepetersen Ok I fixed the default parameters, thanks for catching that.

@miri64 I am still not quite getting what you are saying, sorry for being dense. There is a doc block that opens on Line 33 and closes on Line 56 in params.h, but I believe that was there when you did the review, is there a different kind of doc I should do? I copied the syntax from the hdc1000 params.h file.

miri64 commented on drivers/mma7660/include/mma7660_params.h:
```diff
@@ -0,0 +1,75 @@
+#endif
+
+/**
+ * @brief   Set default configuration parameters for MMA7660 devices
```
This one needs to be `@name` though, otherwise doxygen will add this as doc for just `MMA7660_PARAM_I2C`.

immesys replied:
Wait, does that mean that the [HDC1000](https://github.com/RIOT-OS/RIOT/blob/master/drivers/hdc1000/include/hdc1000_params.h#L30-L47), [LIS3DH](https://github.com/RIOT-OS/RIOT/blob/master/drivers/lis3dh/include/lis3dh_params.h#L30-L59), [mma8x6x](https://github.com/RIOT-OS/RIOT/blob/master/drivers/mma8x5x/include/mma8x5x_params.h#L30-L61) etc etc are all wrong?

immesys replied:
Full disclosure: I know next to nothing about doxygen

miri64 replied:
1. Yes
2. It seems like the params file are not included into [doxygens compilation](http://doc.riot-os.org/files.html) :/

miri64 replied:
Ah, maybe because of [this line](https://github.com/RIOT-OS/RIOT/blob/master/doc/doxygen/riot.doxyfile#L821). So, do whatever you want here as long as people understand it :D

smlng commented on drivers/include/mma7660.h:
```diff
@@ -0,0 +1,238 @@
+{
+  i2c_t i2c;      /**< the I2C handle */
+  uint8_t addr;   /**< the device I2C address */
+} mma7660_t;
```
I think this should rather be:

```
typedef struct {
    mma7660_params_t params;
} mma7660_t;
```

and consequently defined after `mma7660_params_t`. First, both attributes are part of params (so duplicated here), and second you may want to and access the params somewhere, see other drivers like `mag3110`, too.

smlng commented on drivers/mma7660/mma7660.c:
```diff
@@ -0,0 +1,210 @@
+    }
+    i2c_release(dev->i2c);
+
+    if (mma7660_set_mode(dev, 0, 0, 0, 0)) return -1;
```
adapt to RIOT coding style here and lines below.

smlng commented on drivers/mma7660/mma7660.c:
```diff
@@ -0,0 +1,210 @@
+
+int mma7660_read_counts(mma7660_t *dev, int8_t *x, int8_t *y, int8_t *z)
+{
+    int retries = 6;
```
avoid magic numbers, should be configurable via `#define`

smlng commented on drivers/mma7660/mma7660.c:
```diff
@@ -0,0 +1,210 @@
+            i2c_release(dev->i2c);
+            return -1;
+        }
+        if ((t & ALERT_MASK) == 0)
```
coding style, use `{`, `}` even for one-liner `if`

smlng commented on drivers/mma7660/mma7660.c:
```diff
@@ -0,0 +1,210 @@
+            i2c_release(dev->i2c);
+            return -1;
+        }
+        if ((t & ALERT_MASK) == 0)
```
dito, see above

smlng commented on drivers/mma7660/mma7660.c:
```diff
@@ -0,0 +1,210 @@
+    if (t & SIGN_BIT) t |= SIGN_EXTEND;
+        *x = (int8_t)t;
+
+    while (retries > 0) {
```
note: you haven't reset retries, is that intended?

aabadie commented:
@haukepetersen, @miri64, @smlng do you still see remaining things here?

aabadie commented:
@immesys it would be great if you could test the actual status of this PR and confirm that it works. You can try the `tests/driver_mma7660` test application and the `examples/default`  with `mma7660` module loaded (to have the `saul` command).

smlng commented on drivers/mma7660/mma7660.c:
```diff
@@ -0,0 +1,246 @@
+    i2c_acquire(DEV_I2C);
+
+    while (retries > 0) {
+        if (i2c_read_reg(DEV_I2C, DEV_ADDR, MMA7660_XOUT, &t) != 1) {
```
from the data sheet I see that x|y|zout are placed in row, so IMHO it would make sense to have 

```
uint8_t tmp[3];
...
 if (i2c_read_regs(DEV_I2C, DEV_ADDR, MMA7660_XOUT, &tmp[0], 3) != 3) {
...
```

or not,

immesys replied:
I didn't make this change yet because I vaguely remember having a reason for doing it separately and I don't have the time to do in depth testing again. Is it important?

smlng replied:
at least its not blocking from my side, but for sake of simplicity and efficiency we (you? 😉) should look into this later on.

immesys replied:
Agreed

smlng commented:
apart from my comment above, it looks good.

aabadie commented on drivers/mma7660/include/mma7660_reg.h:
```diff
@@ -0,0 +1,52 @@
+}
+#endif
+
+#endif /* MMA8X5X_REG_H */
```
should be `MMA7660_H` in the comment (probably my fault). That's one of the reason why travis is failaing.

aabadie commented:
please squash (as reported by the CI). I would even recommend to split in 3 commits:
* `drivers/mma7660: initial implementation`: with only the code in `drivers`
* `tests/driver_mma7660: add test application`
* `sys/auto_init/saul: add auto init for mma7660 sensor`

smlng commented:
@haukepetersen and @miri64 please check if you comments are addressed ASAP, so this can go into release 2018.01. 

In the mean time I'll make some hardware tests, found one device in my pocket 😉

smlng commented:
to clarify: init of device works, if I disconnect SDA for instance, I get an error as expected. However, `mma7760_read` gives me all zeros ... Just tested with another board (remote-revb), same results.

tcschmidt commented:
@immesys when will you have time to fix the functional problems reported by @smlng ?

tcschmidt commented:
ping @immesys

smlng commented:
I'll try to have another look on this. I'll have the sensor lying around somewhere, maybe I find the (output) bug. But this also needs adaption to new I2C API, too

smlng commented:
I adopted this PR in #9331 using the new I2C API, too. 

Hence closing this one, hope @immesys doesn't mind.
